### DISCUSSION TITLE
Wast encoder

### DISCUSSION THREAD

**Discussion by ahuoguo** [bytecodealliance/wasm-tools#1883] [Q&A]
Hi! I'm trying to get rid of the named labels, custom sections, inline export, and folded exprs for all the spec test files as our interpreter only deals with plain wat files with numbers. It looks like the `wast` crate already provides support for parsing and a representation for `Wast` files, but there's only encoding to binary for `QuoteWat`. It will be ideal if the `Wast` can be prettyprinted in this case but it doesn't seem to be supported.

Related Issue https://github.com/bytecodealliance/wasm-tools/issues/327

**Comment by alexcrichton** [comment]
Would something like `wasm-tools json-from-wast` fit your use case? That converts to `*.wasm` but that could be converted back to text with `wasm-tools print`. Although now that I type this out I realize that `wasm-tools print` will still print `$foo` names, but a flag could possibly be added to skip that. 

Or perhaps put another way: would you prefer to keep parsing the text format as opposed to parsing the binary format? If you'd prefer binary then `wasm-tools json-from-wast` might be a good fit?

**Reply by ahuoguo** [reply]
We thought about that. This means we should write a parse the JSON file, load the corresponding `wasm`, and perform the test.

We thought it would be better if we can just manipulate the wast file directly. Like in the wasm reference interpreter, it has `./wasm -d memory.wast -o memory.bin.wast` that encodes the module in place into its binary format (but we want to replace the the binary format to plain wat).

**Reply by alexcrichton** [reply]
That's true yeah, using `json-from-wast` would require parsing JSON. 

Otherwise though there's no equivalent to what the reference is doing here in the `wast` crate. If you'd like though I don't think it would be unreasonable to add support for that to the subcommands of `wasm-tools`. For example it might be reasonable to add `*.wast` support to `wasm-tools print` and then add a flag to print either binary modules (e.g. `*.bin.wast`) or avoid printing all names (which I think is what you want here?)

**Reply by ahuoguo** [reply]
Yep! That's what we have in mind.

But to get `wasm-tools print`working on wast. There should be some pretty printing on the `WastDirective` struct which I think only supports parsing for now. This should work somewhat like `JsonBuilder` in `json-from-wast`.

**Reply by alexcrichton** [reply]
I'm not sure where this logic would necessarily best live, but I think starting in `print.rs` for the `print` subcommand might be reasonable? That could do some detection if the input is actually `*.wast` and delegate internally to a different printing routine to implement the pretty-printing. That would complicate the parsing from what it currently is, but that's ok.
